shared/lib_battery_dispatch_automatic_btm.cpp

```diff
@@ -0,0 +1,868 @@
+        battReplacementCostPerkWh, battCycleCostChoice, battCycleCost)
+{
+	_P_target_month = -1e16;
+	_P_target_current = -1e16;
+	_P_target_use.reserve(_num_steps);
+	_P_battery_use.reserve(_num_steps);
+
+	grid.reserve(_num_steps);
+	sorted_grid.reserve(_num_steps);
+
+	for (size_t ii = 0; ii != _num_steps; ii++)
+	{
+		grid.push_back(grid_point(0., 0, 0));
+		sorted_grid.push_back(grid[ii]);
+	}
+
+    if (util_rate != nullptr)
+    {
+        rate = std::shared_ptr<rate_data>(new rate_data(*util_rate));
+    }
+
+    costToCycle();
+}
+
+void dispatch_automatic_behind_the_meter_t::init_with_pointer(const dispatch_automatic_behind_the_meter_t* tmp)
+{
+	_P_target_input = tmp->_P_target_input;
+	_P_target_month = tmp->_P_target_month;
+	_P_target_current = tmp->_P_target_current;
+	grid = tmp->grid;
+
+	// time series data which could be slow to copy. Since this doesn't change, should probably make const and have copy point to common memory
+	_P_load_ac = tmp->_P_load_ac;
+	_P_target_use = tmp->_P_target_use;
+	sorted_grid = tmp->sorted_grid;
+
+    if (tmp->rate != nullptr)
+    {
+        rate = std::shared_ptr<rate_data>(new rate_data(*tmp->rate));
+        rate_forecast = std::shared_ptr<UtilityRateForecast>(new UtilityRateForecast(*tmp->rate_forecast));
+    }
+}
+
+// deep copy from dispatch to thisq
+dispatch_automatic_behind_the_meter_t::dispatch_automatic_behind_the_meter_t(const dispatch_t & dispatch) :
+dispatch_automatic_t(dispatch)
+{
+	const dispatch_automatic_behind_the_meter_t * tmp = dynamic_cast<const dispatch_automatic_behind_the_meter_t *>(&dispatch);
+	init_with_pointer(tmp);
+}
+
+// shallow copy from dispatch to this
+void dispatch_automatic_behind_the_meter_t::copy(const dispatch_t * dispatch)
+{
+	dispatch_automatic_t::copy(dispatch);
+	const dispatch_automatic_behind_the_meter_t * tmp = dynamic_cast<const dispatch_automatic_behind_the_meter_t *>(dispatch);
+	init_with_pointer(tmp);
+}
+
+void dispatch_automatic_behind_the_meter_t::dispatch(size_t year,
+	size_t hour_of_year,
+	size_t step)
+{
+    curr_year = year;
+	size_t step_per_hour = (size_t)(1 / _dt_hour);
+	size_t lifetimeIndex = util::lifetimeIndex(year, hour_of_year, step, step_per_hour);
+
+	update_dispatch(year, hour_of_year, step, lifetimeIndex);
+	dispatch_automatic_t::dispatch(year, hour_of_year, step);
+    if (rate_forecast != nullptr)
+    {
+        std::vector<double> actual_dispatch = { m_batteryPower->powerGrid };
+        rate_forecast->forecastCost(actual_dispatch, year, hour_of_year, step);
+    }
+}
+
+void dispatch_automatic_behind_the_meter_t::update_load_data(std::vector<double> P_load_ac){ _P_load_ac = P_load_ac; }
+void dispatch_automatic_behind_the_meter_t::set_target_power(std::vector<double> P_target){ _P_target_input = P_target; }
+double dispatch_automatic_behind_the_meter_t::power_grid_target() { return _P_target_current; };
+
+void dispatch_automatic_behind_the_meter_t::setup_rate_forecast()
+{
+    if (rate != NULL)
```

Comment:
Looks like this is will be the first exception in the dispatch code - is including ` "../ssc/core.h"` appropriate for throwing `exec_error` or is there another non-crashing exception I could use?

Reply:
You bring up a good point that the code for propagating exceptions as error messages is not in shared. I would guess that you already have checked that `rate` is correctly set up in `battstor` before calling `setup_rate_forecast`, but if you haven't I would put it there first and do the `exec_error`. 

Then, I think it makes sense to get all the error checks done during initialization, which it looks like you kind of did here: https://github.com/NREL/ssc/blob/3163967cf8704009561e0e5b12067d514445a5ea/shared/lib_battery_dispatch_automatic_btm.cpp#L75

So then the question for me is whether or not it makes sense for this dispatch class to be initialized without a rate class? If this class REQUIRES it, I'd throw the error during initialization in a try-catch block so the error can then get rethrown as an `exec_error` and then `dispatch_automatic_behind_the_meter_t` never has to check if `rate` again. If the class doesn't require rate, then is there a clearer way to emphasize why not? Is it only required if you're using FORECAST vs the other dispatch algorithms? If so, I would key off that rather than `rate != NULL`.

Reply:
Thanks. I changed this to checking the dispatch mode instead of whether or not rate is null.